### ISSUE TITLE
Fix Assembly::prepare() brain damage

### ISSUE THREAD

**Issue by friedmud** [idaholab/moose#7569]
### Description of the enhancement or error report
`Assembly::prepare()` needs to get split into two different methods: `prepareResidual()` and `prepareJacobian()`.  Right now we're doing a ridiculous amount of unnecessary work while computing the residual (especially with problems with lots of variables in it).

See here: https://github.com/idaholab/moose/blob/1e635526216d40ec9b0ad94fe1d328d3c620d9a3/framework/src/base/Assembly.C#L948

### Rationale for the enhancement or information for reproducing the error
Remove unnecessary work during residual computation.

### Identified impact
Change to internal APIs

**Comment by permcody** [comment]
I split this up in a local branch. It fails just 2 tests in all of test and modules. The bad news is that I couldn't find any case where it helped. I had @YaqiWang provide me with a highly coupled test with 480 variables and it didn't improve the speed at all. Perhaps I'm just doing something wrong but I don't know if this is really as big of a cost as you might think. I can push my branch if you are interested.

**Comment by friedmud** [comment]
Yeah - I didn't expect incredible speed up... it was just an inefficiency
(and, honestly, a design problem... we're essentially getting unintended
side effects) I noticed in the middle of the array variable stuff.

The fact that tests are failing is a big flag that this needed to get
sorted out!

That said - there should be some speedup. I can try rebasing my array
variable stuff on your branch to see if it helps there. Most likely the
inefficiency of everything else we're currently doing wrong in the many
variable case is swamping this brain damage ;-)

Derek

**Comment by permcody** [comment]
OK - Here's my branch if you want to snag it:
https://github.com/permcody/moose/commit/136f0ce096ab89d1b78a1c5a0d24ae450c88bd25
